### PR TITLE
 BUG: 404 returned when accessing some links in slsa_provenance.json

### PR DESCRIPTION
**Pull request by Amalraj-Joseph-IBM**
https://github.com/WASdev/slsa-maven-plugin/issues/11

### CODE REVIEW

File: src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java [hunk 1]
```diff
@@ -47,15 +49,46 @@ public MavenArtifactResourceDescriptor(Dependency artifact) {
...
+        this.uri = resourceURIGenerator(groupId, artifactId, version);
         setAnnotations();
     }
 
+    private String resourceURIGenerator(String groupId, String artifactId, String version){
+        try{
+            return String.format(URI_FORMAT, groupId.replace(".","/"), artifactId, version);
+        }
+        catch (NullPointerException e){
+            // Group ID is empty — this is an empty dependency, returning null.
```
Comment (jacobwdv) [inline src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java]: would a sub-module dependency be an empty dependency? Is that the only time we anticipate groupID being empty?

File: src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java [hunk 2]
```diff
@@ -47,15 +49,46 @@ public MavenArtifactResourceDescriptor(Dependency artifact) {
...
+
     private void setAnnotations() {
         JsonObjectBuilder annotationsBuilder = Json.createObjectBuilder();
         Utils.addIfNonNullAndNotEmpty(type, "type", annotationsBuilder);
         Utils.addIfNonNullAndNotEmpty(scope, "scope", annotationsBuilder);
         annotations = annotationsBuilder.build();
     }
 
+    @Override
+    public boolean equals(Object obj) {
```
Comment (jacobwdv) [inline src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java:74]: where are the new equals / hashCode() methods used? I don't see any new references. 

I could just be missing something, or is ths just for completeness? Perhaps `isSubModuleDependency` was supposed to use this?

File: src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java (hunk 1 above)
Comment (jacobwdv) [reply src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java]: How would we handle this before our change? I guess it would throw a IllegalFormatException?

**Review by jacobwdv — COMMENTED** [review]
A few questions

File: src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java [hunk 3]
```diff
@@ -47,15 +49,46 @@ public MavenArtifactResourceDescriptor(Dependency artifact) {
...
         setAnnotations();
     }
 
+    private String resourceURIGenerator(String groupId, String artifactId, String version){
+        try{
+            return String.format(URI_FORMAT, groupId.replace(".","/"), artifactId, version);
+        }
+        catch (NullPointerException e){
+            // Group ID is empty — this is an empty dependency, returning null.
+            // Note: There are tests in the slsa-maven-plugin repo with empty dependencies.
```
Comment (jacobwdv) [inline src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java]: I don't think we should reference the slsa-maven-plugin tests in this repo since it is technically independent from this one. We should add a small unit test for this in this project. 

also, which test specifically are you referring to in the slsa-maven-plugin?

File: src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java (hunk 1 above)
Comment (Amalraj-Joseph-IBM) [reply src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java]: Now that we are using Maven Central URIs instead of the Sonatype URIs, I had to make some changes here. This is due to the fundamental difference between the two URI templates.

For Sonatype, the URI would look something like this:
`https://central.sonatype.com/artifact/group.id/artifact.id/version`

But for the same artifact, the Maven Central URI would look like:
`https://repo1.maven.org/maven2/group/id/artifact.id/version`

Dots (`.`) in the group ID are replaced with forward slashes (`/`), but this transformation does not apply to the artifact ID or version (Ref: https://maven.apache.org/repository/layout.html).

So, to format this properly, I had to use `String.replace()` on the group ID — which would throw a `NullPointerException` if the string is empty.

In my experience, we cannot have a Maven artifact with any of the coordinates left empty. However, a [test case](https://github.com/WASdev/slsa-maven-plugin/blob/d1eee61fbe93b1b9fbc295625e603e8111d6022f/src/test/java/com/ibm/slsa/maven/plugin/utils/maven/MavenUtilsTest.java#L110) in the `slsa-maven-plugin` repo invokes a situation like this. The test seems to be a bad_path test, but leaving the group ID empty causes an exception. That’s why I wrapped the logic in a try-catch block.

I can remove the comment I added earlier if it’s not appropriate here.

File: src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java (hunk 3 above)
Comment (Amalraj-Joseph-IBM) [reply src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java]: Now that we are using Maven Central URIs instead of the Sonatype URIs, I had to make some changes here. This is due to the fundamental difference between the two URI templates.

For Sonatype, the URI would look something like this:
`https://central.sonatype.com/artifact/group.id/artifact.id/version`

But for the same artifact, the Maven Central URI would look like:
`https://repo1.maven.org/maven2/group/id/artifact.id/version`

Dots (`.`) in the group ID are replaced with forward slashes (`/`), but this transformation does not apply to the artifact ID or version (Ref: https://maven.apache.org/repository/layout.html).

So, to format this properly, I had to use `String.replace()` on the group ID — which would throw a `NullPointerException` if the string is empty.

In my experience, we cannot have a Maven artifact with any of the coordinates left empty. However, a [test case](https://github.com/WASdev/slsa-maven-plugin/blob/d1eee61fbe93b1b9fbc295625e603e8111d6022f/src/test/java/com/ibm/slsa/maven/plugin/utils/maven/MavenUtilsTest.java#L110) in the `slsa-maven-plugin` repo invokes a situation like this. The test seems to be a bad_path test, but leaving the group ID empty causes an exception. That’s why I wrapped the logic in a try-catch block.

I can remove the comment I added earlier if it’s not appropriate here.

File: src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java (hunk 2 above)
Comment (Amalraj-Joseph-IBM) [reply src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java:74]: These methods are used to implement the logic that removes duplicate entries from the provenance file.
 In the addMavenProjectDependencies method, I added a check to see if the artifact is already being added. That check uses `HashSet.contains()` to determine if the artifact is already present. And Java collections require `hashCode()` and `equals()`

File: src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java (hunk 1 above)
Comment (jacobwdv) [reply src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java]: let's remove the comment and add a small unit test for the empty dependency in this project. I didn't think we could have a Maven artifact with any coordinates left empty either, so this is likely an edge case / not even possible.

File: src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java (hunk 2 above)
Comment (jacobwdv) [reply src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java:74]: right makes sense.

File: src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java (hunk 1 above)
Comment (Amalraj-Joseph-IBM) [reply src/main/java/com/ibm/intoto/attestation/custom/resource/descriptors/maven/MavenArtifactResourceDescriptor.java]: Done

File: pom.xml [hunk 4]
```diff
@@ -5,7 +5,7 @@
 
     <groupId>com.ibm.websphere.appserver.features</groupId>
     <artifactId>intoto4j</artifactId>
-    <version>3.0.1-SNAPSHOT</version>
```
Comment (jacobwdv) [inline pom.xml]: Can we leave this at 3.0.1-SNAPSHOT and i'll release it myself and then bump the version up afterwards?

**Review by jacobwdv — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 4 above)
Comment (Amalraj-Joseph-IBM) [reply pom.xml]: Done

**Review by jacobwdv — APPROVED** [review]
LGTM
